File: src/my_project/interviews/top_150_questions_round_22/ex_88_course_schedule.py
from typing import List, Union, Collection, Mapping, Optional
from abc import ABC, abstractmethod
from collections import deque, defaultdict

class Solution:
    def canFinish(self, numCourses: int, 
                        prerequisites: List[List[int]]) -> bool:
        # Build adjacency list and in-degree array
        graph = defaultdict(list)
        in_degree = [0] * numCourses
        
        for course, prereq in prerequisites:
            graph[prereq].append(course)
            in_degree[course] += 1
        
        # Add all courses with no prerequisites to queue
        queue = deque([i for i in range(numCourses) if in_degree[i] == 0])
        courses_taken = 0
        
        # Process courses
        while queue:
            course = queue.popleft()
            courses_taken += 1
            
            # Reduce in-degree for dependent courses
            for next_course in graph[course]:
                in_degree[next_course] -= 1
                if in_degree[next_course] == 0:
                    queue.append(next_course)
        
        # If we took all courses, no cycle exists
        return courses_taken == numCourses

File: src/my_project/interviews_typescript/top_150_questions_round_1/ex_88_course_schedule.ts
function canFinish(numCourses: number, prerequisites: number[][]): boolean {
    // Build adjacency list and in-degree array
    const graph: Map<number, number[]> = new Map();
    const inDegree: number[] = new Array(numCourses).fill(0);
    
    // Initialize graph
    for (let i = 0; i < numCourses; i++) {
        graph.set(i, []);
    }
    
    // Build graph and calculate in-degrees
    for (const [course, prereq] of prerequisites) {
        graph.get(prereq)!.push(course);
        inDegree[course]++;
    }
    
    // Add all courses with no prerequisites to queue
    const queue: number[] = [];
    for (let i = 0; i < numCourses; i++) {
        if (inDegree[i] === 0) {
            queue.push(i);
        }
    }
    
    let coursesTaken = 0;
    
    // Process courses using BFS (Kahn's algorithm)
    while (queue.length > 0) {
        const course = queue.shift()!;
        coursesTaken++;
        
        // Reduce in-degree for dependent courses
        for (const nextCourse of graph.get(course)!) {
            inDegree[nextCourse]--;
            if (inDegree[nextCourse] === 0) {
                queue.push(nextCourse);
            }
        }
    }
    
    // If we took all courses, no cycle exists
    return coursesTaken === numCourses;
}

